torch_xla/csrc/ops/einsum_backward.cpp

```diff
@@ -0,0 +1,134 @@
+
+#include "tensorflow/compiler/xla/client/lib/matrix.h"
+#include "torch_xla/csrc/helpers.h"
+#include "torch_xla/csrc/lowering_context.h"
+#include "torch_xla/csrc/ops/einsum.h"
+#include "torch_xla/csrc/ops/einsum_utilities.h"
+#include "torch_xla/csrc/ops/infer_output_shape.h"
+#include "torch_xla/csrc/ops/xla_ops.h"
+
+namespace torch_xla {
+namespace {
+
+std::vector<torch::lazy::Value> GetOperandList(
+    c10::ArrayRef<torch::lazy::Value> operands,
+    const torch::lazy::Value& grad_output) {
+  std::vector<torch::lazy::Value> operand_list(operands.begin(),
+                                               operands.end());
+  operand_list.insert(operand_list.begin(), grad_output);
+  return operand_list;
+}
+
+std::vector<xla::XlaOp> BuildEinsumBackward(const xla::XlaOp& grad_output,
+                                            absl::Span<const xla::XlaOp> inputs,
+                                            const std::string& equation) {
+  std::vector<xla::XlaOp> ops;
+  if (inputs.size() == 1) {
+    std::string backward_equation =
+        EinsumUtilities::BuildBackwardsEquation(equation);
+    ops.push_back(xla::Einsum(grad_output, backward_equation));
+  } else if (inputs.size() == 2) {
+    std::vector<std::string> equations =
+        EinsumUtilities::BuildBackwardsEquations(equation);
+
+    xla::PrimitiveType type = XlaHelpers::PromoteType(
+        XlaHelpers::TypeOfXlaOp(grad_output),
+        XlaHelpers::TypeOfXlaOp(inputs[0]), XlaHelpers::TypeOfXlaOp(inputs[1]));
+
+    ops.push_back(xla::Einsum(
+        grad_output, inputs[1], equations[0],
+        xla::PrecisionConfig::Precision::PrecisionConfig_Precision_DEFAULT,
+        type));
+
+    ops.push_back(xla::Einsum(
+        inputs[0], grad_output, equations[1],
+        xla::PrecisionConfig::Precision::PrecisionConfig_Precision_DEFAULT,
+        type));
+  }
+
+  return ops;
+}
+
+xla::Shape NodeOutputShapes(const torch::lazy::Value& grad_output,
+                            const torch::lazy::OpList& inputs,
+                            const std::string& equation) {
+  auto lower_for_shape_one_fn =
+      [&](absl::Span<const xla::XlaOp> operands) -> xla::XlaOp {
+    return BuildEinsumBackward(
+        operands[0],
+        std::vector<xla::XlaOp>(operands.begin() + 1, operands.end()),
+        equation)[0];
+  };
+
+  auto lower_for_shape_two_fn =
+      [&](absl::Span<const xla::XlaOp> operands) -> xla::XlaOp {
+    return BuildEinsumBackward(
+        operands[0],
+        std::vector<xla::XlaOp>(operands.begin() + 1, operands.end()),
+        equation)[1];
+  };
+
+  std::vector<xla::Shape> input_shapes;
+  input_shapes.push_back(GetXlaShape(grad_output));
+  for (auto const& op : inputs) {
+    input_shapes.push_back(GetXlaShape(op));
+  }
+
+  xla::Shape output_shape;
+  auto shape_one =
+      InferOutputShape(absl::MakeSpan(input_shapes), lower_for_shape_one_fn);
+
+  if (inputs.size() != 1) {
+    auto shape_two =
```

Comment:
move the `lower_for_shape_two_fn` to here maybe?

Reply:
No longer needed